File: python/be/contract/chain/abi.py
"""A small ABI coder: enough of the Solidity ABI to talk to PerpDEX.

Mirror of ``node/be/contract/chain/abi.js``; verified against the same
``cast``-produced vectors in ``test_abi.py``. Both stacks must produce the
same bytes for the same call, and both are checked against a third
implementation, so a disagreement is never a tie.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from be.contract.chain.keccak import keccak256, selector


@dataclass
class T:
    kind: str
    bits: int = 256
    n: int = 0
    inner: "T | None" = None
    length: int | None = None
    components: list = field(default_factory=list)


def parse_type(s: str) -> T:
    s = s.strip()
    m = re.match(r"^(.*)\[(\d*)\]$", s)
    if m:
        return T("array", inner=parse_type(m.group(1)), length=int(m.group(2)) if m.group(2) else None)
    if s.startswith("("):
        if not s.endswith(")"):
            raise ValueError("unbalanced tuple: " + s)
        return T("tuple", components=[parse_type(p) for p in _split_top(s[1:-1])])
    m = re.match(r"^uint(\d*)$", s)
    if m:
        return T("uint", bits=int(m.group(1)) if m.group(1) else 256)
    m = re.match(r"^int(\d*)$", s)
    if m:
        return T("int", bits=int(m.group(1)) if m.group(1) else 256)
    if s == "address":
        return T("address")
    if s == "bool":
        return T("bool")
    m = re.match(r"^bytes(\d+)$", s)
    if m:
        return T("bytesN", n=int(m.group(1)))
    if s == "bytes":
        return T("bytes")
    if s == "string":
        return T("string")
    raise ValueError("unsupported type: " + s)


def _split_top(s: str) -> list[str]:
    out, depth, start = [], 0, 0
    for i, ch in enumerate(s):
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        elif ch == "," and depth == 0:
            out.append(s[start:i])
            start = i + 1
    if s.strip():
        out.append(s[start:])
    return out


def is_dynamic(t: T) -> bool:
    if t.kind in ("string", "bytes"):
        return True
    if t.kind == "array":
        return t.length is None or is_dynamic(t.inner)
    if t.kind == "tuple":
        return any(is_dynamic(c) for c in t.components)
    return False


def static_size(t: T) -> int:
    if t.kind == "array":
        return t.length * static_size(t.inner)
    if t.kind == "tuple":
        return sum(static_size(c) for c in t.components)
    return 32


MASK = (1 << 256) - 1


def word(n: int) -> str:
    return format(n & MASK, "064x")


def encode_value(t: T, v) -> str:
    if t.kind == "uint":
        n = int(v)
        if n < 0 or n >= 1 << t.bits:
            raise ValueError(f"value {v} does not fit uint{t.bits}")
        return word(n)
    if t.kind == "int":
        n = int(v)
        lim = 1 << (t.bits - 1)
        if n < -lim or n >= lim:
            raise ValueError(f"value {v} does not fit int{t.bits}")
        return word(n)
    if t.kind == "address":
        return str(v).lower().replace("0x", "").rjust(64, "0")
    if t.kind == "bool":
        return word(1 if v else 0)
    if t.kind == "bytesN":
        return str(v).replace("0x", "").ljust(64, "0")
    if t.kind in ("string", "bytes"):
        hx = str(v).encode("utf-8").hex() if t.kind == "string" else str(v).replace("0x", "")
        ln = len(hx) // 2
        return word(ln) + hx.ljust(((ln + 31) // 32) * 64, "0")
    if t.kind == "array":
        items = list(v)
        if t.length is not None and len(items) != t.length:
            raise ValueError("fixed array length mismatch")
        body = encode_tuple([t.inner] * len(items), items)
        return (word(len(items)) + body) if t.length is None else body
    if t.kind == "tuple":
        vals = list(v) if isinstance(v, (list, tuple)) else [v[i] for i in range(len(t.components))]
        return encode_tuple(t.components, vals)
    raise ValueError("cannot encode " + t.kind)


def encode_tuple(types: list[T], values: list) -> str:
    if len(types) != len(values):
        raise ValueError(f"expected {len(types)} values, got {len(values)}")
    heads, tails = [], []
    head_size = sum(32 if is_dynamic(t) else static_size(t) for t in types)
    tail_off = head_size
    for t, v in zip(types, values):
        enc = encode_value(t, v)
        if is_dynamic(t):
            heads.append(word(tail_off))
            tails.append(enc)
            tail_off += len(enc) // 2
        else:
            heads.append(enc)
    return "".join(heads) + "".join(tails)


def encode_call(signature: str, args: list) -> str:
    open_ = signature.index("(")
    types = parse_type(signature[open_:]).components
    return selector(signature) + encode_tuple(types, args)  # selector() is 0x-prefixed


def _decode_value(t: T, hx: str, offset: int):
    def at(o: int) -> str:
        return hx[o * 2 : o * 2 + 64]

    if t.kind == "uint":
        return int(at(offset), 16), 32
    if t.kind == "int":
        n = int(at(offset), 16)
        if n >= 1 << 255:
            n -= 1 << 256
        return n, 32
    if t.kind == "address":
        return "0x" + at(offset)[24:], 32
    if t.kind == "bool":
        return int(at(offset), 16) != 0, 32
    if t.kind == "bytesN":
        return "0x" + at(offset)[: t.n * 2], 32
    if t.kind in ("string", "bytes"):
        ln = int(at(offset), 16)
        raw = hx[(offset + 32) * 2 : (offset + 32) * 2 + ln * 2]
        return (bytes.fromhex(raw).decode("utf-8") if t.kind == "string" else "0x" + raw), 32
    if t.kind == "array":
        base, length = offset, t.length
        if length is None:
            length = int(at(offset), 16)
            base = offset + 32
        items, _ = _decode_tuple([t.inner] * length, hx, base)
        return items, (32 if t.length is None else static_size(t))
    if t.kind == "tuple":
        vals, _ = _decode_tuple(t.components, hx, offset)
        return vals, (32 if is_dynamic(t) else static_size(t))
    raise ValueError("cannot decode " + t.kind)


def _decode_tuple(types: list[T], hx: str, base: int):
    out, cursor = [], base
    for t in types:
        if is_dynamic(t):
            rel = int(hx[cursor * 2 : cursor * 2 + 64], 16)
            v, _ = _decode_value(t, hx, base + rel)
            out.append(v)
            cursor += 32
        else:
            v, used = _decode_value(t, hx, cursor)
            out.append(v)
            cursor += used
    return out, cursor - base


def decode(type_string: str, hx: str) -> list:
    t = parse_type(type_string if type_string.startswith("(") else "(" + type_string + ")")
    body = str(hx).replace("0x", "", 1) if str(hx).startswith("0x") else str(hx)
    vals, _ = _decode_tuple(t.components, body, 0)
    return vals


def named(fields: list[str], values: list) -> dict:
    return dict(zip(fields, values))


def topic(signature: str) -> str:
    return "0x" + keccak256(signature.encode("utf-8")).hex()


def error_table(signatures: list[str]) -> dict:
    table = {}
    for sig in signatures:
        open_ = sig.index("(")
        table[selector(sig).replace("0x", "")] = {"name": sig[:open_], "types": sig[open_:]}
    return table


def decode_error(data: str | None, table: dict) -> dict:
    hx = (data or "").replace("0x", "", 1) if (data or "").startswith("0x") else (data or "")
    if len(hx) < 8:
        return {"name": "unknown", "args": [], "raw": data}
    sel = hx[:8]
    entry = table.get(sel)
    if not entry:
        return {"name": "unknown", "args": [], "raw": data, "selector": "0x" + sel}
    args = [] if entry["types"] == "()" else decode(entry["types"], "0x" + hx[8:])
    return {"name": entry["name"], "args": args, "raw": data, "selector": "0x" + sel}

File: services/mini-api/lib/abi.js
'use strict';

// No hashing here: selectors and topics come precomputed from constants.js.
const { SELECTORS } = require('./constants');
function selector(signature) {
  const s = SELECTORS[signature];
  if (!s) throw new Error('no precomputed selector for ' + signature + ' -- add it to constants.js');
  return s;
}

/**
 * A small ABI coder: enough of the Solidity ABI to talk to PerpDEX.
 *
 *   static:  uintN intN address bool bytes32 uint8-enums
 *   dynamic: string bytes T[] and tuples that contain any of those
 *   nested:  tuple(...) to any depth, arrays of tuples
 *
 * A copy of the coder in node/be/contract/chain/abi.js, minus hashing: this
 * service is the object under test and must not import from the test code.
 * The two copies are compared byte-for-byte in the service selftest.
 *
 * Type strings use Solidity's canonical spelling, e.g.
 *   "placeOrder((uint16,uint8,uint8,uint8,uint64,uint64,uint64,uint8,bool,uint32,uint64))"
 *   "((uint64,uint64,uint8)[],(uint64,uint64,uint8)[])"
 */

// ---------------------------------------------------------------- type parsing

function parseType(s) {
  s = s.trim();
  const arr = s.match(/^(.*)\[(\d*)\]$/);
  if (arr) return { kind: 'array', inner: parseType(arr[1]), length: arr[2] === '' ? null : Number(arr[2]) };
  if (s.startsWith('(')) {
    if (!s.endsWith(')')) throw new Error('unbalanced tuple: ' + s);
    return { kind: 'tuple', components: splitTop(s.slice(1, -1)).map(parseType) };
  }
  let m;
  if ((m = s.match(/^uint(\d*)$/))) return { kind: 'uint', bits: m[1] ? Number(m[1]) : 256 };
  if ((m = s.match(/^int(\d*)$/))) return { kind: 'int', bits: m[1] ? Number(m[1]) : 256 };
  if (s === 'address') return { kind: 'address' };
  if (s === 'bool') return { kind: 'bool' };
  if ((m = s.match(/^bytes(\d+)$/))) return { kind: 'bytesN', n: Number(m[1]) };
  if (s === 'bytes') return { kind: 'bytes' };
  if (s === 'string') return { kind: 'string' };
  throw new Error('unsupported type: ' + s);
}

function splitTop(s) {
  const out = [];
  let depth = 0, start = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '(') depth++;
    else if (s[i] === ')') depth--;
    else if (s[i] === ',' && depth === 0) { out.push(s.slice(start, i)); start = i + 1; }
  }
  if (s.trim() !== '') out.push(s.slice(start));
  return out;
}

function isDynamic(t) {
  if (t.kind === 'string' || t.kind === 'bytes') return true;
  if (t.kind === 'array') return t.length === null || isDynamic(t.inner);
  if (t.kind === 'tuple') return t.components.some(isDynamic);
  return false;
}

/** Size in bytes of the head of a static type. */
function staticSize(t) {
  if (t.kind === 'array') return t.length * staticSize(t.inner);
  if (t.kind === 'tuple') return t.components.reduce((n, c) => n + staticSize(c), 0);
  return 32;
}

// ---------------------------------------------------------------- encoding

const MASK256 = (1n << 256n) - 1n;

function word(bi) {
  return ((BigInt(bi) & MASK256).toString(16)).padStart(64, '0');
}

function encodeValue(t, v) {
  switch (t.kind) {
    case 'uint': {
      const n = BigInt(v);
      if (n < 0n || n >= (1n << BigInt(t.bits))) throw new Error(`value ${v} does not fit uint${t.bits}`);
      return word(n);
    }
    case 'int': {
      const n = BigInt(v);
      const lim = 1n << BigInt(t.bits - 1);
      if (n < -lim || n >= lim) throw new Error(`value ${v} does not fit int${t.bits}`);
      return word(n < 0n ? (1n << 256n) + n : n);
    }
    case 'address': return String(v).toLowerCase().replace(/^0x/, '').padStart(64, '0');
    case 'bool': return word(v ? 1n : 0n);
    case 'bytesN': return String(v).replace(/^0x/, '').padEnd(64, '0');
    case 'string':
    case 'bytes': {
      const hex = t.kind === 'string' ? Buffer.from(String(v), 'utf8').toString('hex') : String(v).replace(/^0x/, '');
      const len = hex.length / 2;
      return word(len) + hex.padEnd(Math.ceil(len / 32) * 64, '0');
    }
    case 'array': {
      const items = Array.from(v);
      if (t.length !== null && items.length !== t.length) throw new Error('fixed array length mismatch');
      const body = encodeTuple(items.map(() => t.inner), items);
      return t.length === null ? word(items.length) + body : body;
    }
    case 'tuple': return encodeTuple(t.components, Array.isArray(v) ? v : t.components.map((_, i) => v[i]));
    default: throw new Error('cannot encode ' + t.kind);
  }
}

function encodeTuple(types, values) {
  if (types.length !== values.length) throw new Error(`expected ${types.length} values, got ${values.length}`);
  const heads = [], tails = [];
  let headSize = types.reduce((n, t) => n + (isDynamic(t) ? 32 : staticSize(t)), 0);
  let tailOffset = headSize;
  types.forEach((t, i) => {
    const enc = encodeValue(t, values[i]);
    if (isDynamic(t)) {
      heads.push(word(tailOffset));
      tails.push(enc);
      tailOffset += enc.length / 2;
    } else {
      heads.push(enc);
    }
  });
  return heads.join('') + tails.join('');
}

/** Full calldata for `signature` with `args`. */
function encodeCall(signature, args) {
  const open = signature.indexOf('(');
  const types = parseType(signature.slice(open)).components;
  return selector(signature) + encodeTuple(types, args);
}

// ---------------------------------------------------------------- decoding

function decodeValue(t, hex, offset) {
  // returns [value, bytesConsumedInHead]
  const at = (o) => hex.slice(o * 2, o * 2 + 64);
  switch (t.kind) {
    case 'uint': return [BigInt('0x' + at(offset)), 32];
    case 'int': {
      let n = BigInt('0x' + at(offset));
      if (n >= 1n << 255n) n -= 1n << 256n;
      return [n, 32];
    }
    case 'address': return ['0x' + at(offset).slice(24), 32];
    case 'bool': return [BigInt('0x' + at(offset)) !== 0n, 32];
    case 'bytesN': return ['0x' + at(offset).slice(0, t.n * 2), 32];
    case 'string':
    case 'bytes': {
      const len = Number(BigInt('0x' + at(offset)));
      const raw = hex.slice((offset + 32) * 2, (offset + 32) * 2 + len * 2);
      return [t.kind === 'string' ? Buffer.from(raw, 'hex').toString('utf8') : '0x' + raw, 32];
    }
    case 'array': {
      let base = offset, length = t.length;
      if (length === null) { length = Number(BigInt('0x' + at(offset))); base = offset + 32; }
      const [items] = decodeTuple(Array.from({ length }, () => t.inner), hex, base);
      return [items, t.length === null ? 32 : staticSize(t)];
    }
    case 'tuple': {
      const [vals] = decodeTuple(t.components, hex, offset);
      return [vals, isDynamic(t) ? 32 : staticSize(t)];
    }
    default: throw new Error('cannot decode ' + t.kind);
  }
}

function decodeTuple(types, hex, base) {
  const out = [];
  let cursor = base;
  for (const t of types) {
    if (isDynamic(t)) {
      const rel = Number(BigInt('0x' + hex.slice(cursor * 2, cursor * 2 + 64)));
      const [v] = decodeValue(t, hex, base + rel);
      out.push(v);
      cursor += 32;
    } else {
      const [v, used] = decodeValue(t, hex, cursor);
      out.push(v);
      cursor += used;
    }
  }
  return [out, cursor - base];
}

/** Decode return data / event data against a tuple type string like "(uint64,uint8)". */
function decode(typeString, hex) {
  const t = parseType(typeString.startsWith('(') ? typeString : '(' + typeString + ')');
  const body = String(hex).replace(/^0x/, '');
  const [vals] = decodeTuple(t.components, body, 0);
  return vals;
}

/** Name the fields of a decoded tuple. */
function named(fields, values) {
  const o = {};
  fields.forEach((f, i) => { o[f] = values[i]; });
  return o;
}

/** Decode a custom error `0x........<args>` given a table {selector: {name, types}}. */
function decodeError(data, table) {
  const hex = String(data || '').replace(/^0x/, '');
  if (hex.length < 8) return { name: 'unknown', args: [], raw: data };
  const sel = hex.slice(0, 8);
  const entry = table[sel];
  if (!entry) return { name: 'unknown', args: [], raw: data, selector: '0x' + sel };
  const args = entry.types === '()' ? [] : decode(entry.types, '0x' + hex.slice(8));
  return { name: entry.name, args, raw: data, selector: '0x' + sel };
}

/** Build a selector table from error signatures like "InsufficientCollateral(int256,uint256)". */
function errorTable(signatures) {
  const table = {};
  for (const sig of signatures) {
    const open = sig.indexOf('(');
    table[selector(sig).replace(/^0x/, '')] = { name: sig.slice(0, open), types: sig.slice(open) };
  }
  return table;
}

module.exports = { parseType, isDynamic, encodeCall, encodeTuple, encodeValue, decode, named, decodeError, errorTable, word };
